test/libyul/YulPureInterpreterTest.cpp

```diff
@@ -0,0 +1,188 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <test/libyul/YulPureInterpreterTest.h>
+
+#include <libyul/tools/interpreter/PureInterpreter.h>
+
+#include <test/Common.h>
+
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/YulStack.h>
+#include <libyul/AsmAnalysisInfo.h>
+#include <libyul/AST.h>
+
+#include <liblangutil/DebugInfoSelection.h>
+#include <liblangutil/ErrorReporter.h>
+#include <liblangutil/SourceReferenceFormatter.h>
+
+#include <libsolutil/AnsiColorized.h>
+#include <libsolutil/Visitor.h>
+
+#include <boost/test/unit_test.hpp>
+#include <boost/algorithm/string.hpp>
+
+using namespace solidity;
+using namespace solidity::util;
+using namespace solidity::langutil;
+using namespace solidity::yul;
+using namespace solidity::yul::test;
+using namespace solidity::frontend;
+using namespace solidity::frontend::test;
+
+using solidity::util::h256;
+
+using namespace solidity::yul::tools::interpreter;
+
+YulPureInterpreterTest::YulPureInterpreterTest(std::string const& _filename):
+	EVMVersionRestrictedTestCase(_filename)
+{
+	m_source = m_reader.source();
+	m_expectation = m_reader.simpleExpectations();
+
+	m_config.maxTraceSize = m_reader.sizetSetting("maxTraceSize", 128);
+	m_config.maxExprNesting = m_reader.sizetSetting("maxExprNesting", 64);
+	m_config.maxSteps = m_reader.sizetSetting("maxSteps", 512);
+	m_config.maxRecursionDepth = m_reader.sizetSetting("maxRecursionDepth", 64);
+
+	m_printHex = m_reader.boolSetting("printHex", false);
```

Comment:
> For testing, the existing `YulInterpreterTest` was not used. This is because this pure interpreter has no memory/storage access, it can not use the same trace as the current interpreter from 'test/libyul/tools'.

That's actually easy to solve. Just add a setting you can set to mark a test as one that requires access to memory or storage. Then add it to those tests that do not pass without it. As you can see in the snippet above, adding settings is pretty easy.

For this we could call the setting `sideEffectFree` and it `true` by default.

Reply:
I understand how to mark a test that requires memory access. However, I’d still insist to keep the two test suites separate. While they do share the parsing part, the settings and trace printing are quite different. I’m concerned that merging the two might make it harder to manage the tests.

I also believe the two interpreter implementations might grow in different directions, so separating the test suites makes more sense to me.

Reply:
While we're ok with leaving the interpreters separate for now to avoid bringing over unnecessary baggage from the test interpreter, the test suite should definitely be shared. Even if the implementations diverge significantly, they ultimately both use Yul as their interface so it should not matter for testing. The same set of tests should be usable for both and I strongly think this is the point we should be starting from. Seeing the new interpreter pass the existing tests (at least the ones it can run) will be an additional confirmation that it works correctly.

I also don't see why settings and trace would have to be substantially different. If the new ones are better, we should just change them in the old test suite as well. I'd suggest actually extracting that into a PR of its own, which we'd merge first.

Reply:
I agree that the testing interface is the same. I will adjust the test suit to use the existing one.

The settings is also not that different. The only notable settings that the pure interpreter has is the max recursion depth. I decided to add this, because this interpreter will be used while walking the AST. This allows the code to be interpreted with a big `maxStep`, while stack memory is still protected from overflow.

The trace is definitely different. The trace of the interpreter in `test/tools` consists the following:
- Storage layout
- Memory layout
- Transient layout
- Traces about memory access, log

The pure interpreter does __not__ evaluate instructions that alter memory/storage, or produces log. Therefore such trace can not be created. I have changed the trace in to the call trace and the values of the outer most variables.

The new trace can be included into the `test/tools` interpreter, but not the other way around.

Reply:
I have another suggestion. I will merge the two testsuit together. For each test, I will print the traces from both interpreter.
How does this sound?

Reply:
> The only notable settings that the pure interpreter has is the *max recursion depth*.

Do we need it as a test setting though? I'm assuming that this will be a hard-coded limit, normally, so why use a different one in tests?

> The pure interpreter does **not** evaluate instructions that alter memory/storage, or produces log. Therefore such trace can not be created.

We can make the pure interpreter always print those the same way the current interpreter does when it's running on pure code. Or we could make the existing interpreter not show this when the traces are empty. This does not seem like a significant obstacle to unifying them.

> I have another suggestion. I will merge the two testsuit together. For each test, I will print the traces from both interpreter.
> How does this sound?

But one will always be empty so why even print it? I think that having the existing interpreter only print the non-empty trace would be the better way to handle this.

Reply:
> I'm assuming that this will be a hard-coded limit in the compiler, so why use a different one in tests?

I forgot that we indeed hardcode the limit in the compiler. In the test, I tried to make it as flexible as possible. But since we'll hardcode anyway, I'll remove settings and test with a known limit.

> We can make the pure interpreter always print those the same way the current interpreter does when it's running on pure code. 

I still dont see the point of this. The trace of the current interpreter is all about state changes. The pure interpreter does not touch the state, so there is no such trace to print.

> Or we could make the existing interpreter not show this when the traces are empty.

In case of empty trace, the current interpreter does have zero trace about state change. But by default, the pure interpreter also have no trace about state change as well. That's why I added another kind of trace so we can see if it works correctly.

>> For each test, I will print the traces from both interpreter.
>
> But one will always be empty so why even print it?

Not exactly.
- For the case the code has state change, the pure interpreter will still have the call trace up to the point where the state change happens, and the `ImpureInstructionEncountered` will be reported.
- For the case where there is no state change, the current interpreter will not have any trace. But we know that for this test, the current interpreter will still do its job.

My idea for the suggestion to print both traces is to test both interpreter at the same test. Both interpreter can run the same yul code, just with different terminated status and traces. And I think this work wrll for the PR: it will use the same testsuit, but it does not need to touch the code for the current interpreter.

Reply:
Ah, so it looks like your trace and the trace in the current interpreter are different things. I assumed they were the same and that your trace was just a subset of the old one. Looks like instead you're recording function calls, while the current interpreter records builtins?

Another thing is that I assumed that the trace was a feature of the test case, not the interpreter itself. It's a debug feature, and this made sense in an interpreter built for testing, but for executing *correct* code in the optimizer that's pure overhead. I think that this should be done by having the base interpreter provide virtual functions as hooks, and having a derived test-only variant that stores the trace.

Overall, there's no technical reason why both interpreters couldn't provide the same output and IMO the best solution would be to just extend the current interpreter to report function calls too. But ok, that's extra work, which I didn't expect this would require so I guess just printing whichever trace is available is acceptable for this PR. We should give them different names though, because they're really different things.

Reply:
Thank you for understanding!

> ... so I guess just printing one kind in one interpreter and the other kind in the other is acceptable for this PR

With this, I'll merge the 2 test-suits into one, and print both traces for the same test case.

> We should give them different names though, because they're really different things.

I am not sure whose _name_ are you referring to. You mean the 2 interpreters should have different names, or their traces should have different names?

> ... IMO the best solution would be to just extend the current interpreter to report function calls too

That what I have originally done in #15358 :pray: . I moved the interpreter from `test/tools` into `libyul` and have the pure interpreter extending that interpreter. I did not change the trace though but that can also be done.

> ... but for executing correct code in the optimizer that's pure overhead

I am positive that my way of printing trace do not have overhead. Here is [godbolt the demo for my trace][trace demo]. If we hardcode `maxTraceSize` to 0, compiler should eliminate all `addTrace` calls.

[trace demo]: https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIMwAcpK4AMngMmAByPgBGmMQgAdIADqgKhE4MHt6%2B/kGp6Y4CYRHRLHEJSbaY9kUMQgRMxATZPn6B1bWZDU0EJVGx8YnSCo3Nrbkdo739ZRXDAJS2qF7EyOwc5gDM4cjeWADUJltuAG6YDiTH2CYaAILbu/uYRydOo8SYrNe3D2Y7DD2XkOxzOTTwhgIP3uj0Bz1eHmSdTE0Ie9wA9OiDvxiAcsCwBB8mHUDgJaABPX5edJGA4AMS8gLqABFMDQGBkBK9mQdRugQCAPuFgMcrPdqcKDl4zABWABs3IOtAEwCVKtFv019z2TAUCgOABViEx1mEWIQAEpfZAITDoTUAdjFDwdzI1MI9dw%2BXgc9MZDkybjEtCNJswjudtwAnAymZlWezOQwDmgGKNzAqsIm6u60VG%2BQLzpdiKDpfLrgdkk1WApc1qY/66kHaCHjesIPXo7GAwIE%2BEkynCVCzAqAPpZ/t1UidjT5gj8kBFoglk5luUV1PpkcHUdV40sWseqMLEAzqMTjl1CDjtmTzJLM97mvXp8Hhadp0mV2Ot1bSP3b1fW7OorQIVYGFDdYI3rYD41vS9MkHNNh0zeCkzrWcC0XC5l1LWV1y2bADk%2BMDiAidAADUxC8TBaz/LU81ggRQPAyDMA7I8u0bODs0QzcUJ3C8k2nTjMPnQscKuVd8I3IcMx3EjwLtKjvFo%2BsTzPISrxvXiBAfUSo0UsjlOo2jryM8iVJohR304z9vxhb96M9CVaRCVBgDY7BBGIclFSw05wUhUEmKyYM2NIP042YzBSIgttw0IjCvQIYgfQIA5lFWTAAElBHiZISPiDwGBoEVHP/KNMV5WKDiIA4NDq1A8TwBQmBieg6rbYV63SAAvTBRwylgmFUNihDwAbFQ0ZLo36wbhtGhpMGSfVjh5GbnLzeahoOEbVGwVRCsiWjHFpdaGow/NJoWvbRqtZBVnSXsVoIBBprrJz/zRFK0t9LLPjyggCqK4gemB6CjwB3L8uIQrYuKgQyqQ0YkLK2bMJu3aGB8ZbVo%2Bra5vE7Di1BdzPIS7zUspQiurDNath5L8rC%2Bhjo0xLE7nQdADiYNVVVSsMADp6w5g4AHU6FoPFmoYVA3slPAqDRvBgCF/bxpu7kLpmo8OejABaOrbX5g5GGpg4AHcpYOOWMriFHUvSu0DnCY3WpTXVwyPYGWGSAwIZOAhyWSRhWBeNiDRDzBIuD0PmDYIWk4OO5iGAOibiPLDUCRTIURONizUta1bXtWmmG5tiIFT9PHYzJOhZ3Jp09svNmbPJWDggVMyvV0bNam9adYWYjYvA3liZxlsc6hQnZ2jTuIEF9YFCF%2BaIBH2TStVvuxoSiaBtb6Nj9nCzDQSouCCtE1S43q7o2X2ihcwP2DHWUcYhNABrCAsPCXc34LWjqCSOwDCKRVHM3VeSd3xzyjNGM%2BWEp7KiRPfByP1oynFQHgHm6AfDJDYgoX%2BxM0ipS%2BCweSo4VgEBHvxT6v5vrhEWuEDeRxPz3AOJwqUNJVQJ1olWdYE8y7fS4ZlbKQMQYI2ICVZGPdVa5lEXItWGt95ax1qKTh1VQQGx0bovReiUwIEMMAF4Ct9T1SUKYk2bIqA4QYqI6GEi4ag3Bi8aYEMnRKK/Aw%2BxXChS0ioMmC65gzD8FQCEhRXClxSTcGuCsr4GZMydFwSKZhIpbG8ZEzh0SVyxJkrTCyJlVKJLYRYaQBwZSRXXKzDhfjGjAyFhXdAbEQrcQEM2VsYZrjUAYJFBJsDnSiPcZgRplcEqtOigwFiZEWmER6ZFQplFTI2WSqInEXcmBeHqnQ7cj8QAT2JCMx%2BNlSm/FEaIxemyiAACpsRhRbIqExBBRxKwmT2e5nSoJzIzI/BYm92F3HOUCwcWyESgjuR0g21wtKIVBOCkJnsWwu2tm9I4ZgzBgpOBC4MULCIJLXjdVhcKsUItfFI1eCLiVuDNgwdAtAsnnO8WbWgShXbKwgFc1AtzAnTKCYzA4TyXlUDeSBMeMzxnfJHL8/5gzgWKOoZi6lPKxUMFxbgNCJIqVooxYsq2hB3qUpOOC5VcU1WLKsk/dem8jUkvRaPOKLtArFJFnarVrh6VbTlZk2pnDvUumZBwJYtBOAyl4H4bgvBUCcDcNYawvIVhrBeNsHgpACCaEDUsL%2BIAZQyiFoEAIUYAgygCP8LYWw5RRgdPoTgkheAsAkBoDQpBw1aFIFGjgvAFAgCbWmjgWglhwFgDARAKBUB%2BzoPEcglA0DjvoAkYAUg0k0FoMDYgXaIAxHTaQGI4Qmjkk4CmndzBfIAHkYjaBwge3gM62CCBPQwCkW6sAxC8MADpXaI2kHxMY8QfbeD4E%2BAGc4H7W2YFUBcLZGwU1MJqFe/QeAYjGl8h4LAcHUp4HrX%2B0g5xiAxDSJgVkI0jC0GFOmpYVADDpwongTAlsT3xzg/wQQIgxDsCkDIQQigVDqCw7oFJBgjAoFjZYeDMQu2QCWDPTIH6DZ8nWqYSw1guAOgOAbE9ZhI04eIDg2i8Alh2BwpkFwtKJh%2BBSaEcIAxyhDBSQUJMpm9B2bqLMQYCQUkGfeT0cYng2h6A83ULzfRLNzBs7YMYLQfO5Hc%2BFlz1m3P6YTesCQQaQ1hq3e2g4qgAhygNnKSQArkDIAOFIPNXdcCEBIGirYXAFi8F7f2pYtoK5DA3qQLNZgox5qjJIDQkgzAOhlBoOUcouBcDlFsatHBa2kHrQ6B0QsNBcF60NgIo2HTVeq829LnBO3dtTWR0gg6R3UOSFsqd3cx3JAncQSI4cQAhEOaMYI%2BBlw4L0Ex4QohxDsY%2B1xtQW6%2BOkEtsaZIV6UscFDVtrD7aT1bNOxlVAysss5bywVorJWMUQA8LO%2BIVWat1bI41r4WAEiteDVNut/gzBC0LVGGUUZqsBBLXKbLRaoetvbbtnthO2sgEkJIIWDpRuLckMp4b43euTa2Gl6HO39t/r%2BZNjTM2G1NpbZGuX9WM3YfiM9PwkggA%3D%3D%3D

Reply:
> I am not sure whose *name* are you referring to. You mean the 2 interpreters should have different names, or their traces should have different names?

Traces. Just make it clear that one shows the function call stack while the other one is the record of evaluated builtins.

> That what I have originally done in #15358 🙏.

Yeah, and that still would be an option if you want to do it. But I don't want to add too much extra work here. Just printing both things is fine for now too.

> If we hardcode `maxTraceSize` to 0, compiler should eliminate all `addTrace` calls.

That's good, bit I'd still rather have the trace functionality only in the test case. The less complexity in the actual implementation, the better.

Reply:
Anyway, my suggestion would be to submit your test cases and the changes in the `YulInterpreterTest` as a separate PR so that we can review and merge that first. They should already all pass on the existing interpreter. Then we can continue with this PR.